### DISCUSSION TITLE
Transcoding for Philips Streamiums

### DISCUSSION THREAD

**Discussion by nobanzai** [gerbera/gerbera#3294] [General]
Hi *,

because my Philips Streamiums won't play 24bit FLACs, I tried to let gerbera transcode these files, but didn't succeed.
Transcoding doesn't seem to be started, because there are no debugging entries in the log.
Any idea, what might be wrong here? This is the excerpt from my config containing the (hopefully) relevant parts:
[config-part.xml.txt](https://github.com/user-attachments/files/18043942/config-part.xml.txt)

TIA.

Bye.
Michael.

**Comment by KarlStraussberger** [comment]
You should check you client configuration on the clients page to make sure that the NP is detected. You may also configure the client by IP - sometimes the devices send differnt string on browse and discovery.

If that is all fine you can try with `<hide-original-resource>yes</hide-original-resource>` because the second resource may confuse the player. Also you could change the mimetype to `<mimetype>audio/x-flac<mimetype>`

**Reply by nobanzai** [reply]
I modified the the mimetype and the client definition:
`   <client
      userAgent="Allegro-Software-WebClient"
      flags="TRANSCODE1"
    />
`
seems to be ok - gerbera now starts the TRANSCODE1 profile, when the Streamium tries to play a song.
I don't get any sound, though, no matter if I set <hide-original-resource> to yes or not.
Sometimes the client even hangs. I have to unplug the power cord then to get it back.

**Reply by nobanzai** [reply]
I get

`2024-12-07 13:12:51.311  debug: [transcode_ext_handler.cc:103] serveContent(): Running profile command: 'ffmpeg', arguments: '-loglevel error -i /home/archiv/sound/Michaels_Musik/Bowie,_David/2002-Heathen/01-Sunday.flac -sample_fmt s16 -ar 44100 /tmp/grb-tr-8f972347-ef9e-4f07-9049-26c94cd3bcf7'
`

with this profile now:

```
     <profile name="flac24to16" enabled="yes" type="external">
        <mimetype>audio/x-flac</mimetype>
        <dlna-profile>FLAC</dlna-profile>
        <accept-url>no</accept-url>
        <first-resource>yes</first-resource>
        <hide-original-resource>yes</hide-original-resource>
        <accept-ogg-theora>no</accept-ogg-theora>
        <sample-frequency>44100</sample-frequency>
        <audio-channels>2</audio-channels>
        <agent command="ffmpeg"
               arguments="-loglevel error -i %in -sample_fmt s16 -ar 44100 %out"
        />
        <thumbnail>no</thumbnail>
        <buffer size="1048576" chunk-size="131072" fill-size="262144" />
      </profile>
```

The Streamiums shows "buffering" forever then.

**Reply by udippel** [reply]
I'd try something like this:
```
<profile name="audio2mp3" enabled="yes" type="external">
        <mimetype>audio/mpeg</mimetype>
        <accept-url>no</accept-url>
        <first-resource>no</first-resource>
        <hide-original-resource>no</hide-original-resource>
        <accept-ogg-theora>no</accept-ogg-theora>
        <agent command="ffmpeg" arguments="-y -i %in -f mp3 %out" />
        <buffer size="1048576" chunk-size="131072" fill-size="262144" />
      </profile>
```
with .flac instead, or likewise. That is starting with a very basic setting, if you haven't done that yet.

Not because I'd expect it to work, but to see that transcoding is done properly. With a client like BubbleUPnP you should be able to see and listen to the two streams. Serves no further purpose but to show that transcoding is actually working. 
From here you could go and start tweaking the parameters. I did likewise, because gerbera's transcoding is unforgiving.

**Reply by nobanzai** [reply]
After migrating from mediatomb to gerbera I played around a lot with transoding and got some very complicated profiles working. But that was a lot of gerbera versions in the past.
Even those formerly working profiles won't work with the actual version. I am not sure, why.
As for bubbleupnp - I can't use it, because there is no Android device in my internal network. WLAN is available only in our DMZ, which does not allow to route multicast traffic to the internal network. You may call it paranoia.

But I got it working with the following profile now:

```
      <transcode
        mimetype="audio/x-flac"
        client-flags="TRANSCODE1"
        using="mh-flac2mp3"
      />
 ```

```
     <profile name="mh-flac2mp3" enabled="yes" type="external">
        <mimetype>audio/mpeg</mimetype>
        <dlna-profile>MP3</dlna-profile>
        <accept-url>no</accept-url>
        <first-resource>yes</first-resource>
        <hide-original-resource>no</hide-original-resource>
        <accept-ogg-theora>no</accept-ogg-theora>
        <agent command="ffmpeg" arguments="-y -i %in -f mp3 %out" />
        <buffer size="1048576" chunk-size="131072" fill-size="262144" />
      </profile>

```
Thx for the starting point!

**Reply by nobanzai** [reply]
With this profile, the client hangs forever with message "Buffering":

```
      <profile name="mh-flac24to16" enabled="yes" type="external">
        <mimetype>audio/x-flac</mimetype>
        <dlna-profile>FLAC</dlna-profile>
        <accept-url>no</accept-url>
        <first-resource>yes</first-resource>
        <hide-original-resource>no</hide-original-resource>
        <accept-ogg-theora>no</accept-ogg-theora>
        <!-- sample-frequency>44100</sample-frequency -->
        <!-- audio-channels>2</audio-channels -->
        <agent command="ffmpeg"
               arguments="-i %in -sample_fmt s16 -ar 48000 %out"
        />
        <thumbnail>no</thumbnail>
        <buffer size="1048576" chunk-size="131072" fill-size="262144" />
      </profile>
```

**Reply by udippel** [reply]
Again, unfortunately, only a generic answer.
1. Try - if you haven't done it yet - the same command (parameters) file->file. Just to check that your parameters work with ffmpeg as expected.
2. This is relevant, because some (container) formats don't code in flight, bit by bit. They rather refer to different sections of the files, eventually already coded output. That's all fine when handling files, but not when handling data from/to FiFos. Because those will eventually not be available any longer. 
There are video files(!) that play almost immediately, while some don't, but want to be downloaded completely beforehand.

**Reply by nobanzai** [reply]
Regarding topic 1 - I already tried that, it works as expected and the result is a valid 16bit 48 K flac.
As for topic 2 - as to my knowledge this is not true for flac. Am I wrong here?

**Reply by udippel** [reply]
My last effort would be to try
`arguments="-y -i %in -f flac %out" />`
If this doesn't work, I think we can happily retire the idea.

**Reply by nobanzai** [reply]
I added a view manually from 24 to 16 bit converted FLACs to the gerbera database and tried to play them directly from the Streamium menu. They played without any problems. So the transcoding itself doesn't seem to be the problem.
Perhaps the buffer settings?

**Reply by udippel** [reply]
I'm afraid I've lost you here.
If you feed a 16-bit flac to Streamium, it works. That's what we expected from the beginning, didn't we?
```
<first-resource>yes</first-resource>
<hide-original-resource>no</hide-original-resource>
```
It plays the original (manually reduced) 16-bit resource. How would we know that transcoding works, then?

Maybe I only wasn't clever enough to understand, though?

**Reply by nobanzai** [reply]
> My last effort would be to try `arguments="-y -i %in -f flac %out" />` If this doesn't work, I think we can happily retire the idea.

This works for all 16bit FLACs, but not for the 26bit FLACs, because they don't get transcoded then.

**Reply by nobanzai** [reply]
> I'm afraid I've lost you here. If you feed a 16-bit flac to Streamium, it works. That's what we expected from the beginning, didn't we?
> 
> ```
> <first-resource>yes</first-resource>
> <hide-original-resource>no</hide-original-resource>
> ```
> 
> It plays the original (manually reduced) 16-bit resource. How would we know that transcoding works, then?

For this test, I deactivated the transcoding profile. I just wanted to know, if the result of the ffmpeg transcoding can be played at all.

> 
> Maybe I only wasn't clever enough to understand, though?

Probably my English isn't good enough to explain it the right way 8-<
16bit FLACs worked from the beginning - that's correct, but only without transcoding.
As soon as the transcoding profile is activated for FLACs, the "Buffering" problem also happens for 16bit FLACs.
As far as I understand, there is no way to feed only the 24bit FLACs through the transcoding pipe.

**Reply by KarlStraussberger** [reply]
You may be able to restrict transcoding if your playing a bit with `source-profile` (https://docs.gerbera.io/en/latest/config-transcode.html#transcode)

**Reply by nobanzai** [reply]
> You may be able to restrict transcoding if your playing a bit with `source-profile` (https://docs.gerbera.io/en/latest/config-transcode.html#transcode)

Is there any parameter to specify the bit depth?

**Reply by udippel** [reply]
> This works for all 16bit FLACs, but not for the 26bit FLACs, because they don't get transcoded then.

Ah, good, so transcoding to FLAC principally works. 
It is curious that ffmpeg does it properly, but seemingly not when reading/writing to buffers. I wonder why that would be the case?
I don't think that 16 bit can be distinguished from 24, because I think (others might know better!) that the filtration is done on target MIME, not on actual content. 
Just to confirm: 24-bit FLAC transcode and play as mp3?

**Reply by nobanzai** [reply]
> > You may be able to restrict transcoding if your playing a bit with `source-profile` (https://docs.gerbera.io/en/latest/config-transcode.html#transcode)
> 
> Is there any parameter to specify the bit depth?

bitsPerSample I suppose - but how exactly do I have to use it?

**Reply by nobanzai** [reply]
> > This works for all 16bit FLACs, but not for the 26bit FLACs, because they don't get transcoded then.
> 
> Ah, good, so transcoding to FLAC principally works. It is curious that ffmpeg does it properly, but seemingly not when reading/writing to buffers. I wonder why that would be the case? I don't think that 16 bit can be distinguished from 24, because I think (others might know better!) that the filtration is done on target MIME, not on actual content. Just to confirm: 24-bit FLAC transcode and play as mp3?

In general 16 or 24 bit transcoding to mp3 and streaming to the Streamiums works as expected.

**Reply by nobanzai** [reply]
> > > You may be able to restrict transcoding if your playing a bit with `source-profile` (https://docs.gerbera.io/en/latest/config-transcode.html#transcode)
> > 
> > 
> > Is there any parameter to specify the bit depth?
> 
> bitsPerSample I suppose - but how exactly do I have to use it?

Got it:
```
      <contenttype-dlnaprofile>
        ...
        <map from="flac" bitsPerSample="24" to="FLAC24" />
      </contenttype-dlnaprofile>
      ...
      <transcode
        mimetype="audio/x-flac"
        source-profile="FLAC24"
        client-flags="TRANSCODE1"
        using="mh-flac2mp3"
      />
```
That works as expected - only 24bit FLACs go through the transcoding profile.
But is it correct or does it work at random only?

**Reply by KarlStraussberger** [reply]
It's perfectly fine. Was busy in my Christmas bakery to look up the proper settings myself.

**Reply by KarlStraussberger** [reply]
Only problem may be other devices seeing the `FLAC24`.

**Reply by nobanzai** [reply]
> Only problem may be other devices seeing the `FLAC24`.

Shouldn't be a problem when using TRANSCODE1 for Streamiums only, should it?
If it is a problem, how could I avoid it?

**Reply by nobanzai** [reply]
> It's perfectly fine. Was busy in my Christmas bakery to look up the proper settings myself.

Hey, can you mail me some cookies ;)

**Reply by KarlStraussberger** [reply]
> > Only problem may be other devices seeing the `FLAC24`.
> 
> Shouldn't be a problem when using TRANSCODE1 for Streamiums only, should it?
> If it is a problem, how could I avoid it?

Never thought this through with conditional transcoding. You'll see whether some other player rejects those files.

**Reply by nobanzai** [reply]
After testing with the Streamiums I saw, that the profile is used for my Majik DSM, too, though it does not match the client description. It doesn't matter, if I add or remove the FLAC24 settings - in both cases 24bit FLACs are transcoded for the DSM.

**Reply by nobanzai** [reply]
It happens even when using the ip for the client description.
I assume this is a bug?!

**Reply by nobanzai** [reply]
In the Web gui the three clients look like this:
![image](https://github.com/user-attachments/assets/b82372c5-c8a3-4041-a7db-76eac1c2c992)
```
  <clients
    enabled="yes"
    cache-threshold="8760"
    bookmark-offset="10"
  >
    <client
      userAgent="Allegro-Software-RomPlug"
      flags="TRANSCODE1"
    />
    <client
      userAgent="Allegro-Software-WebClient"
      flags="TRANSCODE1"
    />
  </clients>
  ...
      <contenttype-dlnaprofile>
        <map from="flac" to="FLAC" />
        <map from="flac" bitsPerSample="24" to="FLAC24" />
        ...
      </contenttype-dlnaprofile>
  ...
      <transcode
        mimetype="audio/x-flac"
        source-profile="FLAC24"
        client-flags="TRANSCODE1"
        using="mh-flac2mp3"
      />
  ...
      <profile name="mh-flac2mp3" enabled="yes" type="external">
        <mimetype>audio/mpeg</mimetype>
        <dlna-profile>MP3</dlna-profile>
        <accept-url>no</accept-url>
        <first-resource>yes</first-resource>
        <hide-original-resource>no</hide-original-resource>
        <accept-ogg-theora>no</accept-ogg-theora>
        <agent command="ffmpeg" arguments="-y -i %in -f mp3 %out" />
        <buffer size="1048576" chunk-size="131072" fill-size="262144" />
      </profile>
```
Is there anything wrong with the config?
If not, should I file a bug?

**Reply by nobanzai** [reply]
Uhuh, thats really strange. I stopped gerbera as for every config change, but this time it took a bit longer, until I started it again.
And all of a sudden the above mentioned problems were gone - everything seems to work as expected.
I am not sure, what happened. I always checked, if gerbera and all ffmpeg processes had been gone, before modifying the config and starting gerbera again.
Does gerbera keep some information in the database for a certain time? I had a look into the database, bu couldn't find such information there.

**Reply by KarlStraussberger** [reply]
The only thing we keep in the database is the client info which may influence the filtering. I am master in changing the config without restarting. Keeps you busy looking for phantom issues.

**Reply by nobanzai** [reply]
> The only thing we keep in the database is the client info which may influence the filtering. I am master in changing the config without restarting. Keeps you busy looking for phantom issues.

In this case I really made sure, that gerbera was stopped - at least for the second or third attempt.
As for the clients in database: Is it possible, to remove old entries without removing the database in total? Because the Streamiums formerly had a few other ip addresses, they now have multiple entries in the database.

**Reply by nobanzai** [reply]
Transcoding for the Streamiums now works as expected, but I can't stop gerbera from transcoding for the Majik DSM, too - despite the fact, that it is not included in a client definition pointing to the TRANSCODE1 setting. Even if I switch to ips in the client settings, gerbera transcodes for the DSM.
Whats more - the DSM can't even play the transcoded MP3 anymore.
All settings are exactly as shown above. And yes, they also used to work for the Majik DSM.
I am really clueless now 8-<
Is there any chance, that the database entry for the DSM is the reason for this. It does not contain any description fields.

**Comment by nobanzai** [comment]
One more question: When playing a song directly from the Streamium menu, transcoding is initiated, but when playing from a different control point like upplay, no transcoding takes place.
Is that a limitation of upplay, of gerbera or of the UPnP concept as such?

**Reply by nobanzai** [reply]
Any idea about that topic?

**Reply by KarlStraussberger** [reply]
The default for the threshold is 6 hours (https://docs.gerbera.io/en/latest/config-clients.html#clients). After that they are purged.

**Reply by nobanzai** [reply]
Doesn't work, as it seems.
![image](https://github.com/user-attachments/assets/47626231-a0f4-4bdc-b13a-8288b059a65f)
Clients stay in database for a long time (here 23 Nov).
Besides that, names are not updated - there are clients in the database with a name, that has been changed weeks ago.

**Reply by nobanzai** [reply]
But the question in this thread was about the control point and transcoding.
Sorry, this topic gets too long, I know.

**Reply by KarlStraussberger** [reply]
Didn't see the new topic on my mobile :-<

If you're using a control point we do not know for which device the request is. You may check the network requests again to extract the headers sent by the control point.

**Reply by KarlStraussberger** [reply]
> Doesn't work, as it seems. ![image](https://private-user-images.githubusercontent.com/10534549/393815643-47626231-a0f4-4bdc-b13a-8288b059a65f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NzM3NjEsIm5iZiI6MTczMzc3MzQ2MSwicGF0aCI6Ii8xMDUzNDU0OS8zOTM4MTU2NDMtNDc2MjYyMzEtYTBmNC00YmRjLWIxM2EtODI4OGIwNTlhNjVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDE5NDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiM2ZlMzlhZmEwY2VjNTljZmU3MDFhMjc4NTY4Mzk0YjhiM2RlZDI5Mzk2MzQzYjlmYmM1MGU0YWM5MzcxZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1qeFnRcCsBlTtINN6AA36vpQqRMCGvvaajhSXObTq1I) Clients stay in database for a long time (here 23 Nov). Besides that, names are not updated - there are clients in the database with a name, that has been changed weeks ago.

I'll have to investigate this topic

**Reply by nobanzai** [reply]
> Didn't see the new topic on my mobile :-<

no problem :)

> 
> If you're using a control point we do not know for which device the request is. You may check the network requests again to extract the headers sent by the control point.

Because the Majik DSM has no built in control point as the Streamiums, I need to use upplay or any other control point - otherwise I can't stream any songs to the Majik DSM.
Ok, I created a network trace for this scenario, but if I try to upload it, it vanishes immediately without an error message from the posting. I tried all variants.

**Reply by nobanzai** [reply]
I talked to the developer of upplay, sent him the cp protocol and asked him, if upplay and Gerbera are playing together as intended. I got the answer below - you probably know all this already, but the topic regarding the format switch sounds interesting:

> Gerbera does not know what device will be actually using the resources, it only talks to the control point, which itself talks to the renderer. So the only thing Gerbera can do is provide additional resource links with every track, and let the CP choose, according to the renderer.
> 
> Furthermore, I am not aware of renderers sending details beyond the MIME type in the protocolinfo entries, so I don't think that the Streamium is telling what exact FLAC sample rate and width it supports.
> 
> All in all, I think that the only element which can do the right thing here is upplay, but it does not have the capacity to do it at the moment: the format match function would have to be set per renderer.
> 
> There is already a way to specify some renderer-specific parameters, so I could have a look at adding the format match switch there.

**Comment by nobanzai** [comment]
All relevant questions have been discussed and mostly answered, so I close this discussion.
